design/benchmark_tests.md

```diff
@@ -0,0 +1,33 @@
+# Benchmarks Tests
+
+## Requirements
+1. Measure the latency percentiles (50th, 90th, 95th, 99th 99.9) for a batch of repeated basic operations (e.g. 10K put requests),
+   at a predefined parallelism level.
+2. Show trends of benchmarks tests execution times across different versions.
+3. Portability: The tests don't assume anything on the environment where the lakeFS & DB are running.
+   lakeFS address is the only **required** argument for running the tests.
+
+## Non-Requirements
+1. An active gatekeeper - single slow running cycle shouldn't block the build.
+2. Expect 100% reliability of all actions - when performing thousands of operations,
+   some (exact percent TBD) will be allowed to fail. When measuring an external app,
+   it's the app responsibility to retry on transient errors.
+3. Avoid repeating unit/integration/system tests scenarios.
+  
+## Solution
+### Test Execution
+For each parallelism level, the flow will create a repository and will PUT 10k files in it.
+Then, it will try to read those files. Finally, it will delete them.
+
+### CI
+Github workflow is triggered on a merge to master.
```

Comment:
Just a thought: if this also blocks merges to master, maybe it's better to initially do it as a nightly run + report?

Advantage: Does not increase our mean-time-to-merge-to-master which is a metric I just invented
Disadvantage: We might merge something to master that causes a performance degradation. We can still make sure it has to run before creating a release to avoid user impact.

Not sure which is better, simply floating the idea :)

Reply:
I did not plan to have this as a blocker at first.